hail/src/main/scala/is/hail/expr/ir/ExtractIntervalFilters.scala

```diff
@@ -0,0 +1,277 @@
+        val (s2, i2) = processPredicates(x2, kType)
+        (s1.union(s2), Interval.union(i1 ++ i2, kType.ordering.intervalEndpointOrdering))
+
+      case Conjunction(x1, x2) =>
+        val (s1, i1) = processPredicates(x1, kType)
+        val (s2, i2) = processPredicates(x2, kType)
+        log.info(s"intersecting list of ${ i1.length } intervals with list of ${ i2.length } intervals")
+        val intersection = Interval.intersection(i1, i2, kType.ordering.intervalEndpointOrdering)
+        log.info(s"intersect generated ${ intersection.length } intersected intervals")
+        (s1.union(s2), intersection)
+
+      case Unknown => noData // should only be found in a conjunction
+    }
+  }
+
+  def opIsSupported(op: ComparisonOp[_]): Boolean = {
+    op match {
+      case _: Compare => false
+      case _: NEQ => false
+      case _: NEQWithNA => false
+      case _: EQWithNA => false
+      case _ => true
+    }
+  }
+
+  def extract(cond1: IR, ref: Ref, k: IR): KeyFilterPredicate = {
+    cond1 match {
+      case ApplySpecial("||", Seq(l, r)) =>
+        val ll = extract(l, ref, k)
+        val rr = extract(r, ref, k)
+        if (ll == Unknown || rr == Unknown)
+          Unknown
+        else
+          Disjunction(ll, rr)
+      case ApplySpecial("&&", Seq(l, r)) =>
+        Conjunction(extract(l, ref, k), extract(r, ref, k))
+      case Coalesce(Seq(x, False())) => extract(x, ref, k)
+      case x@ApplyIR("contains", Seq(_: Literal, `k`)) => LiteralContains(x) // don't match string contains
+      case x@ApplySpecial("contains", Seq(_: Literal, `k`)) => IntervalContains(x)
+      case x@ApplyIR("contains", Seq(_: Literal, Apply("contig", Seq(`k`)))) => LocusContigContains(x)
+      case x@ApplyComparisonOp(op, l, r) if opIsSupported(op) =>
+        if (IsConstant(l) && r == k || l == k && IsConstant(r))
+          // TODO: need to look for casts, since patterns like [ `k` > 1.5 ] will not match if `k` is an integer
+          KeyComparison(x)
+        else if ((IsConstant(l) && r == Apply("contig", FastSeq(k))
+          || l == Apply("contig", FastSeq(k)) && IsConstant(r)) && op.isInstanceOf[EQ])
+          LocusContigComparison(x)
+        else if (IsConstant(l) && r == Apply("position", FastSeq(k))
+          || l == Apply("position", FastSeq(k)) && IsConstant(r))
+          LocusPositionComparison(x)
+        else
+          Unknown
+      case Let(name, _, body) if name != ref.name =>
+        // TODO: thread key identity through values, since this will break when CSE arrives
+        // TODO: thread predicates in `value` through `body` as a ref
+        extract(body, ref, k)
+      case _ => Unknown
+    }
+  }
+
+  def extractPartitionFilters(cond: IR, ref: Ref, key: IndexedSeq[String]): Option[(Array[Interval], IR)] = {
+    if (key.isEmpty)
+      return None
+
+    val k1 = GetField(ref, key.head)
+
+    val (nodes, intervals) = processPredicates(extract(cond, ref, k1), k1.typ)
```

Comment:
You don't need to make this change, but I think you could refactor this in a way that avoids building an intermediate `KeyFilterPredicate`, but keeping the code mostly unchanged.

For example, in `processPredicates` replace
```
      case KeyComparison(comp) =>
        val (v, isFlipped) = if (IsConstant(comp.l)) (comp.l, false) else (comp.r, true)
        Set[IR](comp) -> Array(openInterval(constValue(v), v.typ, comp.op, isFlipped))
```
by 
```
def keyComparison(comp): (Set[IR], Array[Interval]) = {
    val (v, isFlipped) = if (IsConstant(comp.l)) (comp.l, false) else (comp.r, true)
    Set[IR](comp) -> Array(openInterval(constValue(v), v.typ, comp.op, isFlipped))
}
```
and instead of constructing a `KeyComparison` node in `extract`, just call `keyComparison` directly, combining `processPredicates(extract(cond, ref, k1), k1.typ))` into a single recursive traversal.

Reply:
ah, yeah, you're totally right. I had to do some transformations of the `KeyFilterPredicate` objects in an earlier iteration, but now this should be straightforward! I'll take a stab at it.

Reply:
yep, even simpler after this change!